docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,217 @@
+mempool acts more like a gossip data structure.
+
+The second ABCI method, `ProcessProposal`, is used to process the block proposer's
+proposal as defined by `PrepareProposal`. This ABCI method requests that the
+application evaluate the entire proposed block for validity.
+
+It is important to note that in this phase of ABCI++ integration, the application
+is NOT responsible for locking semantics -- Tendermint will still be responsible
+for that. In the future, however, the application will be responsible for locking,
+which allows for parallel execution possibilities.
+
+## Decision
+
+We will integrate the first phase of ABCI++, which will be introduced in Tendermint
+v0.37.0, in the next major release of the Cosmos SDK. We will integrate the two
+aforementioned ABCI++ methods on the `BaseApp` type. We describe the implementations
+of the two methods individually below.
+
+Prior to describing the implementation of the two new methods, it is important to
+note that the existing ABCI methods, `CheckTx`, `DeliverTx`, etc, still exist and
+serve the same functions as they do now.
+
+### `PrepareProposal`
+
+Prior to evaluating the decision for how to implement `PrepareProposal`, it is
+important to note that `CheckTx` will still be executed and will be responsible
+for evaluating transaction validity as it does now, with one very important
+_additive_ distinction.
+
+When executing transactions in `CheckTx`, the application will now add valid
+transactions, i.e. passing the AnteHandler, to it's own mempool data structure.
+In order to provide a flexible approach to meet the varying needs of application
+developers, we will define both a mempool interface and a data structure utilizing
+Golang generics, allowing developers to focus only on transaction
+ordering. Developers requiring absolute full control can implement their own
+custom mempool implementation.
+
+We define the general mempool interface as follows (subject to change):
+
+```go
+// MempoolTx we define a mempool transaction interface that is as minimal as
+// possible, only requiring applications to define the size of the transaction
+// to be used when reaping and getting the transaction itself. Interface type
+// casting can be used in the actual mempool implementation.
+type MempoolTx interface {
+	// Size returns the size of the transaction in bytes.
+	Size(codec.Codec) int
+	Tx() sdk.Tx
+}
+
+type Mempool[T MempoolTx] interface {
+	// Insert attempts to insert a MempoolTx into the mempool returning an error
+	// upon failure.
+	Insert(T) error
+	// Reap returns the next available transaction in the mempool, returning an error
+	// upon failure. The notion of 'available' or 'next' is defined by the application's
+	// mempool implementation.
+	Reap() (T, error)
+	// ReapMaxBytes performs the the same function as Reap, except it returns
+	// multiple transactions limited by their total size.
+	ReapMaxBytes(n int) ([]T, error)
```

Comment:
ref: https://github.com/cosmos/cosmos-sdk/pull/12838#discussion_r943798638

We can probably remove `Reap` altogether  and modify `ReapMaxBytes` to be `ReapMaxBytes(txRecords tm.TxRecords, maxBytes int64) ([]T, error)